pkg/notifications/notifications.go

```diff
@@ -0,0 +1,494 @@
+	}
+
+	type IcingaDbWebRule struct {
+		Version int `json:"version"` // expect icingaDbWebRuleVersion
+		Config  struct {
+			Type   string `json:"type"`   // expect one of [all, host, service]
+			Filter string `json:"filter"` // Icinga DB Web filter expression
+		} `json:"config"`
+		Queries struct {
+			Host    *IcingaDbWebQuery `json:"host"`
+			Service *IcingaDbWebQuery `json:"service,omitempty"`
+		} `json:"queries"`
+	}
+
+	outRuleIds := make([]string, 0, len(client.rulesInfo.Rules))
+
+	for id, filterExpr := range client.rulesInfo.Rules {
+		if filterExpr == "" {
+			outRuleIds = append(outRuleIds, id)
+			continue
+		}
+
+		var webRule IcingaDbWebRule
+		if err := json.Unmarshal([]byte(filterExpr), &webRule); err != nil {
+			return nil, errors.Wrap(err, "cannot decode rule filter expression as JSON into struct")
+		}
+
+		if version := webRule.Version; version != icingaDbWebRuleVersion {
+			return nil, errors.Errorf("decoded rule filter expression .Version is %d, %d expected", version, icingaDbWebRuleVersion)
+		}
+		if cfgType := webRule.Config.Type; !slices.Contains(
+			[]string{icingaDbWebRuleTypeAll, icingaDbWebRuleTypeHost, icingaDbWebRuleTypeService}, cfgType) {
+			return nil, errors.Errorf("decoded rule filter expression contains unsupported .Config.Type %q", cfgType)
+		}
+		if cfgType := webRule.Config.Type; cfgType != icingaDbWebRuleTypeService && webRule.Queries.Host == nil {
+			return nil, errors.Errorf("decoded rule filter expression for .Config.Type %q with an empty .Queries.Host", cfgType)
+		}
+		if cfgType := webRule.Config.Type; cfgType != icingaDbWebRuleTypeHost && webRule.Queries.Service == nil {
+			return nil, errors.Errorf("decoded rule filter expression for .Config.Type %q with an empty .Queries.Service", cfgType)
+		}
+
+		var webQuery IcingaDbWebQuery
+		if !serviceId.Valid() {
+			if webRule.Config.Type == icingaDbWebRuleTypeService {
+				continue
+			}
+			webQuery = *webRule.Queries.Host
+		} else {
+			if webRule.Config.Type == icingaDbWebRuleTypeHost {
+				continue
+			}
+			webQuery = *webRule.Queries.Service
+		}
+
+		queryArgs := make([]any, 0, len(webQuery.Parameters))
+		for _, param := range webQuery.Parameters {
+			switch param {
+			case ":host_id":
+				queryArgs = append(queryArgs, hostId.String())
+			case ":service_id":
+				if !serviceId.Valid() {
+					return nil, errors.New("host rule filter expression contains :service_id for replacement")
+				}
+				queryArgs = append(queryArgs, serviceId.String())
+			case ":environment_id":
+				queryArgs = append(queryArgs, environmentId.String())
+			default:
+				queryArgs = append(queryArgs, param)
+			}
+		}
+
+		evaluates, err := func() (bool, error) {
+			rows, err := client.db.QueryContext(ctx, client.db.Rebind(webQuery.Query), queryArgs...)
+			if err != nil {
+				return false, err
+			}
+			defer func() { _ = rows.Close() }()
+
+			if !rows.Next() {
+				return false, nil
+			}
+			return true, nil
```

Comment:
Why not just `return rows.Next(), nil`? Feels easier to understand for me as well: the rule matches (maybe a better name than "evaluates") if there a next (i.e. a first) row in the result.

Reply:
Changed.